### PR TITLE
23 feature forward headers

### PR DESCRIPTION
**Pull request by lukasmatusiewicz**
Forward configured headers to the privacyIDEA server.

Triggerchallenge is performed before the first HTTP Servlet Request is sent.
todo: is it possible to fetch headers in triggerchallenge?

### CODE REVIEW

File: privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java [hunk 1]
```diff
@@ -102,6 +109,47 @@ protected void extractChallengeData(@Nonnull PIResponse piResponse)
...
+                Enumeration<String> e = request.getHeaders(headerName);
+                if (e != null)
+                {
+                    while (e.hasMoreElements())
+                    {
+                        headerValues.add(e.nextElement());
+                    }
+                }
+
+                if(!headerValues.isEmpty())
```
Comment (nilsbehlen) [inline privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java:138]: space `if (`

File: privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java [hunk 2]
```diff
@@ -102,6 +109,47 @@ protected void extractChallengeData(@Nonnull PIResponse piResponse)
...
+     * @param request http servlet request
+     * @return headers to forward with their values
+     */
+    protected Map<String, String> getHeadersToForward(HttpServletRequest request)
+    {
+        if (piServerConfigContext.getConfigParams().getForwardHeaders() != null && !piServerConfigContext.getConfigParams().getForwardHeaders().isEmpty())
+        {
+            String cleanHeaders = piServerConfigContext.getConfigParams().getForwardHeaders().replaceAll(" ", "");
+            List<String> headersList = List.of(cleanHeaders.split(","));
+            Map<String, String> headersToForward = new LinkedHashMap<>();
```
Comment (nilsbehlen) [inline privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java:124]: Put this before the `if (piServerconfigcontext...` and return headersToForward at the end of the function.
Meaning you can change this function easily to have only 1 return statement

**Review by nilsbehlen — CHANGES_REQUESTED** [review]

File: privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java (hunk 2 above)
Comment (lukasmatusiewicz) [reply privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java]: fixed: [affa5c4](https://github.com/privacyidea/shibboleth-plugin/pull/26/commits/affa5c4906c254c5bc72620f34c8424fcc800410)

File: privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java (hunk 1 above)
Comment (lukasmatusiewicz) [reply privacyIDEA-impl/src/main/java/org/privacyidea/action/AbstractChallengeResponseAction.java]: fixed: [affa5c4](https://github.com/privacyidea/shibboleth-plugin/pull/26/commits/affa5c4906c254c5bc72620f34c8424fcc800410)

**Review by nilsbehlen — APPROVED** [review]
👍

**Pull request merged**